### PR TITLE
feat!: Remove DOM Parser

### PR DESCRIPTION
**Pull request by dave-nicholas**
## Background: 
The native DOM Parser can perform poorly on some older devices, this approach is faster on newer devices but is considerably better on older devices.
This PR replaces the usage of the DOM Parser for DASH, MSS, VTT and TTML.

The draw back of this approach that it does not include any validation at the cost of better performance.

### CODE REVIEW

**Comment by avelad** [comment]
@dave-nicholas Since it change the returns of some externs, i'll change it to feat!

File: lib/dash/mpd_utils.js [hunk 1]
```diff
@@ -477,20 +485,19 @@ shaka.dash.MpdUtils = class {
           // the element can be changed further.
 
           // Remove the current contents of the node.
-          while (element.childNodes.length) {
-            element.removeChild(element.childNodes[0]);
+          while (element.children.length) {
+            element.children.shift();
```
Comment (absidue) [inline lib/dash/mpd_utils.js:489]: Setting element.children to an empty array is probably more efficient than shifting all the elements.

File: lib/dash/content_protection.js [hunk 2]
```diff
@@ -430,7 +450,7 @@ shaka.dash.ContentProtection = class {
     const data = new Uint8Array([]);
     const psshVersion = 1;
     const pssh =
-         shaka.util.Pssh.createPssh(data, systemId, keyIds, psshVersion);
+        shaka.util.Pssh.createPssh(data, systemId, keyIds, psshVersion);
```
Comment (avelad) [inline lib/dash/content_protection.js:453]: revert it

File: lib/dash/content_protection.js [hunk 3]
```diff
@@ -469,12 +489,12 @@ shaka.dash.ContentProtection = class {
         const proInitData = ContentProtection.getInitDataFromPro_(element);
         let clearKeyInitData = null;
         if (element.schemeUri ===
-              shaka.dash.ContentProtection.ClearKeySchemeUri_) {
```
Comment (avelad) [inline lib/dash/content_protection.js:472]: revert it

File: lib/dash/content_protection.js [hunk 4]
```diff
@@ -469,12 +489,12 @@ shaka.dash.ContentProtection = class {
...
+            shaka.dash.ContentProtection.ClearKeySchemeUri_) {
           clearKeyInitData =
-            ContentProtection.getInitDataClearKey_(element, keyIds);
+              ContentProtection.getInitDataClearKey_(element, keyIds);
```
Comment (avelad) [inline lib/dash/content_protection.js:494]: revert it

File: lib/dash/content_protection.js [hunk 5]
```diff
@@ -469,12 +489,12 @@ shaka.dash.ContentProtection = class {
...
         }
         const initData = element.init || defaultInit || proInitData ||
-          clearKeyInitData;
+            clearKeyInitData;
```
Comment (avelad) [inline lib/dash/content_protection.js:497]: revert it

File: lib/dash/content_protection.js [hunk 6]
```diff
@@ -788,7 +809,7 @@ shaka.dash.ContentProtection.CencNamespaceUri_ = 'urn:mpeg:cenc:2013';
  * @private
  */
 shaka.dash.ContentProtection.ClearKeyNamespaceUri_ =
-  'http://dashif.org/guidelines/clearKey';
+    'http://dashif.org/guidelines/clearKey';
```
Comment (avelad) [inline lib/dash/content_protection.js:812]: revert it

File: lib/dash/content_protection.js [hunk 7]
```diff
@@ -804,4 +825,4 @@ shaka.dash.ContentProtection.ClearKeySchemeUri_ =
  * @private
  */
 shaka.dash.ContentProtection.DashIfNamespaceUri_ =
-  'https://dashif.org/CPS';
+    'https://dashif.org/CPS';
```
Comment (avelad) [inline lib/dash/content_protection.js:828]: revert it

File: lib/dash/dash_parser.js [hunk 8]
```diff
@@ -310,6 +311,7 @@ shaka.dash.DashParser = class {
     // Keep track of how long the longest manifest update took.
     const endTime = Date.now();
     const updateDuration = (endTime - startTime) / 1000.0;
+
```
Comment (avelad) [inline lib/dash/dash_parser.js:314]: revert it

File: lib/dash/mpd_utils.js [hunk 9]
```diff
@@ -503,57 +511,61 @@ shaka.dash.MpdUtils = class {
    * Filter the contents of a node recursively, replacing xlink links
    * with their associated online data.
    *
-   * @param {!Element} element
+   * @param {?shaka.extern.xml.Node} parent
```
Comment (avelad) [inline lib/dash/mpd_utils.js:514]: Parent is not used anywhere in this function, so we need to remove it.

File: lib/text/ttml_text_parser.js [hunk 10]
```diff
@@ -205,13 +203,21 @@ shaka.text.TtmlTextParser = class {
...
-      const span = document.createElement('span');
-      span.textContent = cueNode.textContent;
+      /** @type {shaka.extern.xml.Node} */
+      const span = {
+        tagName: 'span',
+        children: [TXml.getTextContents(cueNode)],
+        attributes: {},
+        parent: parentElement,
+      };
+      // for (const key in cueNode.attributes) {
```
Comment (avelad) [inline lib/text/ttml_text_parser.js:213]: remove it

File: lib/text/ttml_text_parser.js [hunk 11]
```diff
@@ -205,13 +203,21 @@ shaka.text.TtmlTextParser = class {
...
+      //   span.attributes[key] = cueNode.attributes[key];
+      // }
       cueElement = span;
     } else {
-      goog.asserts.assert(cueNode.nodeType == Node.ELEMENT_NODE,
-          'nodeType should be ELEMENT_NODE!');
-      cueElement = /** @type {!Element} */(cueNode);
+      // goog.asserts.assert(cueNode.nodeType == Node.ELEMENT_NODE,
```
Comment (avelad) [inline lib/text/ttml_text_parser.js:218]: remove it

File: lib/text/ttml_text_parser.js [hunk 12]
```diff
@@ -239,27 +245,24 @@ shaka.text.TtmlTextParser = class {
...
     // Parse any nested cues first.
-    const isTextNode = (node) => {
-      return node.nodeType == Node.TEXT_NODE;
-    };
-    const isLeafNode = Array.from(cueElement.childNodes).every(isTextNode);
+    const isLeafNode = cueElement.children.every(TXml.isText);
     const nestedCues = [];
     if (!isLeafNode) {
-      // Otherwise, recurse into the children.  Text nodes will convert into
+      // Otherwise, recurse into the children.  Text nodes will convert in    to
```
Comment (avelad) [inline lib/text/ttml_text_parser.js:263]: too many spaces in `in    to`

File: lib/text/vtt_text_parser.js [hunk 13]
```diff
@@ -463,14 +465,25 @@ shaka.text.VttTextParser = class {
     payload = VttTextParser.replaceKaraokeStylePayload_(payload);
     payload = VttTextParser.replaceVoiceStylePayload_(payload);
     const xmlPayload = '<span>' + payload + '</span>';
-    const element = shaka.util.XmlUtils.parseXmlString(xmlPayload, 'span');
+    let element;
+    try {
+      element = shaka.util.TXml.parseXmlString(xmlPayload, 'span');
+    } catch (e) {
+      shaka.log.warning('cue parse fail: ', e);
+      element = {
```
Comment (avelad) [inline lib/text/vtt_text_parser.js:473]: We don't want an element here because we already have a fallback, so this is not necessary

File: test/dash/dash_parser_manifest_unit.js [hunk 14]
```diff
@@ -910,7 +910,9 @@ describe('DashParser Manifest', () => {
   });
 
   describe('fails for', () => {
-    it('invalid XML', async () => {
+    // The cost of performance with the tXml library means that we don't
+    // get validation.
+    xit('invalid XML', async () => {
```
Comment (avelad) [inline test/dash/dash_parser_manifest_unit.js:915]: Wouldn't it be better to eliminate this test?

File: test/dash/dash_parser_manifest_unit.js [hunk 15]
```diff
@@ -920,7 +922,9 @@ describe('DashParser Manifest', () => {
       await Dash.testFails(source, error);
     });
 
-    it('XML with inner errors', async () => {
+    // The cost of performance with the tXml library means that we don't
+    // get validation.
+    xit('XML with inner errors', async () => {
```
Comment (avelad) [inline test/dash/dash_parser_manifest_unit.js:927]: Wouldn't it be better to eliminate this test?

File: test/dash/mpd_utils_unit.js [hunk 16]
```diff
@@ -541,7 +536,9 @@ describe('MpdUtils', () => {
       await testSucceeds(baseXMLString, desiredXMLString, 3);
     });
 
-    it('fails if loaded file is invalid xml', async () => {
+    // The cost of performance with the tXml library means that we don't
+    // get validation.
+    xit('fails if loaded file is invalid xml', async () => {
```
Comment (avelad) [inline test/dash/mpd_utils_unit.js:541]: Wouldn't it be better to eliminate this test?

File: test/mss/mss_parser_unit.js [hunk 17]
```diff
@@ -85,7 +85,9 @@ describe('MssParser Manifest', () => {
   });
 
   describe('fails for', () => {
-    it('invalid XML', async () => {
+    // The cost of performance with the tXml library means that we don't
+    // get validation.
+    xit('invalid XML', async () => {
```
Comment (avelad) [inline test/mss/mss_parser_unit.js:90]: Wouldn't it be better to eliminate this test?

File: test/mss/mss_parser_unit.js [hunk 18]
```diff
@@ -95,7 +97,9 @@ describe('MssParser Manifest', () => {
       await Mss.testFails(source, error);
     });
 
-    it('XML with inner errors', async () => {
+    // The cost of performance with the tXml library means that we don't
+    // get validation.
+    xit('XML with inner errors', async () => {
```
Comment (avelad) [inline test/mss/mss_parser_unit.js:102]: Wouldn't it be better to eliminate this test?

File: test/text/ttml_text_parser_unit.js [hunk 19]
```diff
@@ -57,32 +57,33 @@ describe('TtmlTextParser', () => {
         '<tt xml:space="default">' + ttBody + '</tt>',
         {periodStart: 0, segmentStart: 60, segmentEnd: 70, vttOffset: 0},
         {startTime: 62.03, endTime: 62.05});
+    // NOTE: This cannot be supported (easily) using the tXml parser.
```
Comment (avelad) [inline test/text/ttml_text_parser_unit.js:60]: What is the explanation for this? Do you know what dash.js does in this case when using txml?

File: test/text/vtt_text_parser_unit.js [hunk 20]
```diff
@@ -827,11 +827,12 @@ describe('VttTextParser', () => {
               },
             ],
           },
-          {
-            startTime: 90,
-            endTime: 100,
-            payload: '<b>Test<i>8</b>',
-          },
+          // NOTE: This is isn't going to work with tXml currently
```
Comment (avelad) [inline test/text/vtt_text_parser_unit.js:830]: I have cloned your branch and uncommented this case and it works correctly, please revert this change.

**Review by avelad — CHANGES_REQUESTED** [review]

File: test/text/ttml_text_parser_unit.js (hunk 19 above)
Comment (avelad) [reply test/text/ttml_text_parser_unit.js:60]: And can we make it similar here?

Comment (dave-nicholas) [reply test/text/ttml_text_parser_unit.js:60]: I could add feature to our implementation of tXml to do something similar to this when parsing tt elements.

Comment (avelad) [reply test/text/ttml_text_parser_unit.js:60]: Please

File: docs/tutorials/upgrade.md [hunk 21]
```diff
@@ -96,9 +96,11 @@ application:
   - Configuration changes:
     - `streaming.forceTransmuxTS` has been renamed to `streaming.forceTransmux`
       (deprecated in v4.3.0)
+    - `manifest.dash.manifestPreprocessor` callback now receives a type of `shaka.externs.xml.Node` instead of `Element`.
```
Comment (avelad) [inline docs/tutorials/upgrade.md:99]: Same for MSS. `manifest.mss.manifestPreprocessor`

File: lib/dash/dash_parser.js [hunk 22]
```diff
@@ -481,41 +484,42 @@ shaka.dash.DashParser = class {
 
     let availabilityTimeOffset = 0;
     if (uriObjs && uriObjs.length) {
-      availabilityTimeOffset = XmlUtils.parseAttr(
-          uriObjs[0], 'availabilityTimeOffset', XmlUtils.parseFloat) || 0;
+      availabilityTimeOffset = TXml.parseAttr(
+          uriObjs[0], 'availabilityTimeOffset',
+          TXml.parseFloat) || 0;
```
Comment (avelad) [inline lib/dash/dash_parser.js:489]: This fit into the previous line

File: lib/media/drm_engine.js [hunk 23]
```diff
@@ -28,7 +28,7 @@ goog.require('shaka.util.StreamUtils');
 goog.require('shaka.util.StringUtils');
 goog.require('shaka.util.Timer');
 goog.require('shaka.util.Uint8ArrayUtils');
-goog.require('shaka.util.XmlUtils');
+goog.require('shaka.util.TXml');
```
Comment (avelad) [inline lib/media/drm_engine.js:31]: please sort alphabetically

File: lib/media/drm_engine.js [hunk 24]
```diff
@@ -1638,24 +1639,28 @@ shaka.media.DrmEngine = class {
...
     for (const header of headers) {
-      const name = header.getElementsByTagName('name')[0];
-      const value = header.getElementsByTagName('value')[0];
+      const name = TXml.getElementsByTagName(header, 'name')[0];
+      const value = TXml.getElementsByTagName(header, 'value')[0];
       goog.asserts.assert(name && value, 'Malformed PlayReady headers!');
-      request.headers[name.textContent] = value.textContent;
+      request.headers[
+          /** @type{string} */(shaka.util.TXml.getTextContents(name))] =
+        /** @type{string} */(shaka.util.TXml.getTextContents(value));
```
Comment (avelad) [inline lib/media/drm_engine.js:1653]: nit: add two spaces in the indentation

File: lib/mss/content_protection.js [hunk 25]
```diff
@@ -12,7 +12,7 @@ goog.require('shaka.util.ManifestParserUtils');
 goog.require('shaka.util.Pssh');
 goog.require('shaka.util.StringUtils');
 goog.require('shaka.util.Uint8ArrayUtils');
-goog.require('shaka.util.XmlUtils');
+goog.require('shaka.util.TXml');
```
Comment (avelad) [inline lib/mss/content_protection.js:15]: please sort alphabetically

File: lib/dash/content_protection.js [hunk 26]
```diff
@@ -14,7 +14,7 @@ goog.require('shaka.util.ManifestParserUtils');
 goog.require('shaka.util.Pssh');
 goog.require('shaka.util.StringUtils');
 goog.require('shaka.util.Uint8ArrayUtils');
-goog.require('shaka.util.XmlUtils');
+goog.require('shaka.util.TXml');
```
Comment (avelad) [inline lib/dash/content_protection.js:17]: please sort alphabetically

File: test/dash/dash_parser_manifest_unit.js [hunk 27]
```diff
@@ -1898,11 +1868,13 @@ describe('DashParser Manifest', () => {
...
     const config = shaka.util.PlayerConfiguration.createDefault().manifest;
     config.dash.manifestPreprocessor = (mpd) => {
-      const selector = 'AdaptationSet[mimeType="text/vtt"';
-      const vttElements = mpd.querySelectorAll(selector);
-      for (const element of vttElements) {
-        element.parentNode.removeChild(element);
-      }
+      /** @type {shaka.extern.xml.Node} */
+      const manifest = /** @type {shaka.extern.xml.Node} */ (
+        /** @type {shaka.extern.xml.Node} */(mpd).children[0]);
```
Comment (avelad) [inline test/dash/dash_parser_manifest_unit.js:1873]: nit: add two spaces in the indentation

File: test/dash/mpd_utils_unit.js [hunk 28]
```diff
@@ -452,11 +452,9 @@ describe('MpdUtils', () => {
                       ' />');
       }
       xmlLines.push('</SegmentTimeline>');
-      const parser = new DOMParser();
-      const xml =
-          parser.parseFromString(xmlLines.join('\n'), 'application/xml');
-      const segmentTimeline = xml.documentElement;
-      console.assert(segmentTimeline);
+      const segmentTimeline = /** @type {shaka.extern.xml.Node} */ (
+        shaka.util.TXml.parseXmlString(xmlLines.join('\n'),
```
Comment (avelad) [inline test/dash/mpd_utils_unit.js:456]: nit: add two spaces in the indentation

File: test/dash/mpd_utils_unit.js [hunk 29]
```diff
@@ -691,8 +674,8 @@ describe('MpdUtils', () => {
       /** @type {!shaka.util.PublicPromise} */
       const continuePromise = fakeNetEngine.delayNextRequest();
 
-      const xml = parser.parseFromString(baseXMLString, 'text/xml')
-          .documentElement;
+      const xml = /** @type {shaka.extern.xml.Node} */ (
+        shaka.util.TXml.parseXmlString(baseXMLString));
```
Comment (avelad) [inline test/dash/mpd_utils_unit.js:678]: nit: add two spaces in the indentation

File: test/dash/mpd_utils_unit.js [hunk 30]
```diff
@@ -730,11 +713,11 @@ describe('MpdUtils', () => {
 
     async function testSucceeds(
         baseXMLString, desiredXMLString, desiredNetCalls) {
-      const desiredXML = parser.parseFromString(desiredXMLString, 'text/xml')
-          .documentElement;
+      const desiredXML = /** @type {shaka.extern.xml.Node} */ (
+        shaka.util.TXml.parseXmlString(desiredXMLString));
```
Comment (avelad) [inline test/dash/mpd_utils_unit.js:717]: nit: add two spaces in the indentation

File: test/dash/mpd_utils_unit.js [hunk 31]
```diff
@@ -785,8 +768,8 @@ describe('MpdUtils', () => {
     }
 
     function testRequest(baseXMLString) {
-      const xml = parser.parseFromString(baseXMLString, 'text/xml')
-          .documentElement;
+      const xml = /** @type {shaka.extern.xml.Node} */ (
+        shaka.util.TXml.parseXmlString(baseXMLString));
```
Comment (avelad) [inline test/dash/mpd_utils_unit.js:772]: nit: add two spaces in the indentation

**Review by avelad — CHANGES_REQUESTED** [review]
LGTM, you just have to resolve some minor issues and I will approve the PR. Thanks!

File: lib/media/drm_engine.js (hunk 24 above)
Comment (dave-nicholas) [reply lib/media/drm_engine.js:1653]: The linter will autofix it to look like this no matter what spacing I put.

File: test/dash/dash_parser_manifest_unit.js (hunk 27 above)
Comment (dave-nicholas) [reply test/dash/dash_parser_manifest_unit.js:1873]: The linter reverts to this spacing unfortunately.

Comment (avelad) [reply test/dash/dash_parser_manifest_unit.js:1873]: ok!

File: lib/media/drm_engine.js (hunk 24 above)
Comment (avelad) [reply lib/media/drm_engine.js:1653]: ok!

File: test/dash/mpd_utils_unit.js (hunk 31 above)
Comment (dave-nicholas) [reply test/dash/mpd_utils_unit.js:772]: The linter reverts to this spacing unfortunately.

File: test/dash/mpd_utils_unit.js (hunk 29 above)
Comment (dave-nicholas) [reply test/dash/mpd_utils_unit.js:678]: The linter reverts to this spacing unfortunately.

File: test/dash/mpd_utils_unit.js (hunk 28 above)
Comment (dave-nicholas) [reply test/dash/mpd_utils_unit.js:456]: The linter reverts to this spacing unfortunately.

**Review by avelad — APPROVED** [review]

File: lib/text/ttml_text_parser.js [hunk 32]
```diff
@@ -239,27 +240,24 @@ shaka.text.TtmlTextParser = class {
...
     // Parse any nested cues first.
-    const isTextNode = (node) => {
-      return node.nodeType == Node.TEXT_NODE;
-    };
-    const isLeafNode = Array.from(cueElement.childNodes).every(isTextNode);
+    const isLeafNode = cueElement.children.every(TXml.isText);
     const nestedCues = [];
     if (!isLeafNode) {
-      // Otherwise, recurse into the children.  Text nodes will convert into
+      // Otherwise, recurse into the children.  Text nodes will convert in to
```
Comment (theodab) [inline lib/text/ttml_text_parser.js:258]: No, I think "into" was correct.
https://www.diffen.com/difference/In_To_vs_Into#:~:text=Into%20and%20%22in%20to%22%20are,are%20not%20a%20combined%20phrase.
"Into indicates movement, action or transformation."

File: lib/util/tXml.js [hunk 33]
```diff
@@ -0,0 +1,707 @@
...
+   * Parse some data
+   * @param {string} schema
+   * @param {string} NS
+   */
+  static setKnownNameSpace(schema, NS) {
+    shaka.util.TXml.knownNameSpaces_.set(schema, NS);
+  }
+
+  /**
+     * parseXML / html into a DOM Object,
```
Comment (theodab) [inline lib/util/tXml.js:71]: The indentation on this is wrong.

File: lib/util/tXml.js [hunk 34]
```diff
@@ -0,0 +1,707 @@
...
+
+goog.require('shaka.util.StringUtils');
+goog.require('shaka.log');
+
+/**
+ * This code is a modified version of the tXml library.
+ *
+ * @author: Tobias Nickel
+ * created: 06.04.2015
+ * https://github.com/TobiasNickel/tXml
```
Comment (theodab) [inline lib/util/tXml.js:11]: I... think this means we need to modify our LICENSE file again?

I wish Joey was back, he knows this license stuff better than I do...

File: lib/util/tXml.js [hunk 35]
```diff
@@ -0,0 +1,707 @@
...
+ */
+
+shaka.util.TXml = class {
+  /**
+     * Parse some data
```
Comment (theodab) [inline lib/util/tXml.js:16]: The indentation on this is wrong.

File: lib/util/tXml.js [hunk 36]
```diff
@@ -0,0 +1,707 @@
...
+   * @param {string} xmlString
+   * @param {string=} expectedRootElemName
+   * @return {shaka.extern.xml.Node | null}
+   */
+  static parseXmlString(xmlString, expectedRootElemName) {
+    const result = shaka.util.TXml.parse(xmlString);
+    if (!expectedRootElemName && result.length) {
+      return result[0];
+    }
+    const rootNode = result.find(
```
Comment (theodab) [inline lib/util/tXml.js:37]: This should all fit into one line.

```js
const rootNode = result.find((n) => n.tagName === expectedRootElemName);
```

File: lib/util/tXml.js [hunk 37]
```diff
@@ -0,0 +1,707 @@
...
+    function parseName() {
+      const start = pos;
+      while (nameSpacer.indexOf(S[pos]) === -1 && S[pos]) {
+        pos++;
+      }
+      return S.slice(start, pos);
+    }
+
+    /**
+       * Parse text in current context
```
Comment (theodab) [inline lib/util/tXml.js:214]: The indentation on this is wrong.

File: lib/util/tXml.js [hunk 38]
```diff
@@ -0,0 +1,707 @@
...
+   * @param {string} name The attribute name.
+   * @param {function(string): (T|null)} parseFunction A function that parses
+   *   the attribute.
+   * @param {(T|null)=} defaultValue The attribute's default value, if not
+   *   specified, the attibute's default value is null.
+   * @return {(T|null)} The parsed attribute on success, or the attribute's
+   *   default value if the attribute does not exist or could not be parsed.
+   * @template T
+   */
+  static parseAttr(
```
Comment (theodab) [inline lib/util/tXml.js:469]: This should easily fit into one line.

File: lib/util/tXml.js [hunk 39]
```diff
@@ -0,0 +1,707 @@
...
+    return shaka.util.TXml.parseXmlString(xmlString, expectedRootElemName);
+  }
+
+  /**
+   * Parse some data
+   * @param {string} xmlString
+   * @param {string=} expectedRootElemName
+   * @return {shaka.extern.xml.Node | null}
+   */
+  static parseXmlString(xmlString, expectedRootElemName) {
```
Comment (theodab) [inline lib/util/tXml.js:32]: conformance.textproto mentions shaka.util.XmlUtils.parseXmlString directly. You should update that to mention shaka.util.TXml.parseXmlString instead.

**Review by theodab — CHANGES_REQUESTED** [review]

File: lib/util/tXml.js (hunk 34 above)
Comment (dave-nicholas) [reply lib/util/tXml.js:11]: Thanks for the review @theodab.
The license of the tXml library is MIT https://github.com/TobiasNickel/tXml/blob/master/LICENSE
I think that means that we just need to add the author to our author file? (I must admit I don't know either :) )

Comment (dave-nicholas) [reply lib/util/tXml.js:11]: Great I will do that now

**Comment by avelad** [comment]
@theodab can you review it again? Thanks!

**Review by theodab — APPROVED** [review]
